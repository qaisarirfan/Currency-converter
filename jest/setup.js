jest.mock('react-native-reanimated', () => {
  const Reanimated = require('react-native-reanimated/mock');

  // The mock for `call` immediately calls the callback which is incorrect
  // So we override it with a no-op
  Reanimated.default.call = () => {};

  return Reanimated;
});

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useFocusEffect: () => jest.fn(),
    useRoute: () => jest.fn(),
    useNavigation: () => ({
      navigation: () => jest.fn(),
      canGoBack: () => jest.fn(),
      goBack: () => jest.fn(),
      push: () => jest.fn(),
    }),
  };
});

jest.mock('react-native-bootsplash', () => ({
  hide: jest.fn(),
  show: jest.fn(),
}));

jest.mock('react-native-localize', () => ({
  getLocales: () => [
    { countryCode: 'GB', languageTag: 'en-GB', languageCode: 'en', isRTL: false },
    { countryCode: 'US', languageTag: 'en-US', languageCode: 'en', isRTL: false },
    { countryCode: 'FR', languageTag: 'fr-FR', languageCode: 'fr', isRTL: false },
  ],

  getNumberFormatSettings: () => ({
    decimalSeparator: '.',
    groupingSeparator: ',',
  }),

  getCalendar: () => 'gregorian', // or "japanese", "buddhist"
  getCountry: () => 'US', // the country code you want
  getCurrencies: () => ['USD', 'EUR'], // can be empty array
  getTemperatureUnit: () => 'celsius', // or "fahrenheit"
  getTimeZone: () => 'Europe/Paris', // the timezone you want
  uses24HourClock: () => true,
  usesMetricSystem: () => true,

  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
}));

jest.mock('react-native-keyboard-aware-scroll-view', () => {
  const KeyboardAwareScrollView = require('react-native').ScrollView;
  return { KeyboardAwareScrollView };
});

jest.mock('react-native-gesture-handler', () => {
  // eslint-disable-next-line global-require
  const View = require('react-native/Libraries/Components/View/View');
  return {
    Swipeable: View,
    DrawerLayout: View,
    State: {},
    ScrollView: View,
    Slider: View,
    Switch: View,
    TextInput: View,
    ToolbarAndroid: View,
    ViewPagerAndroid: View,
    DrawerLayoutAndroid: View,
    WebView: View,
    NativeViewGestureHandler: View,
    TapGestureHandler: View,
    FlingGestureHandler: View,
    ForceTouchGestureHandler: View,
    LongPressGestureHandler: View,
    PanGestureHandler: View,
    PinchGestureHandler: View,
    RotationGestureHandler: View,
    /* Buttons */
    RawButton: View,
    BaseButton: View,
    RectButton: View,
    BorderlessButton: View,
    /* Other */
    FlatList: View,
    gestureHandlerRootHOC: jest.fn(),
    Directions: {},
  };
});

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');
