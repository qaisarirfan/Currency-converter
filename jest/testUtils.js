import React from "react"
import {render} from "react-native-testing-library"
import {Provider} from "react-redux"
import configureStore from "redux-mock-store"
import ThemeContextProvider from "../src/ContextUtils/ThemeContext"
import ConversionContextProvider from "../src/ContextUtils/ConversionContext"
import {initialState as themesState} from "../src/redux/reducers/themes"

const INITIAL_STATE = {
  themes: themesState,
}
const mockStore = configureStore([])
const store = mockStore(INITIAL_STATE)

const reduxRender = (ui, options) =>
  render(ui, {
    wrapper: ({children}) => (
      <Provider store={store}>
        <ThemeContextProvider>
          <ConversionContextProvider>{children}</ConversionContextProvider>
        </ThemeContextProvider>
      </Provider>
    ),
    ...options,
  })

// re-export everything for convenience
export * from "react-native-testing-library"

// override render method
export {reduxRender as render, store}
