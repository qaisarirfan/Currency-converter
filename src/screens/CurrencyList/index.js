import React, {useContext} from "react"
import flow from "lodash/flow"
import get from "lodash/get"
import filter from "lodash/filter"
import {View, FlatList, TouchableOpacity} from "react-native"
import PropTypes from "prop-types"
import {useNavigation, useRoute} from "@react-navigation/native"
import Entypo from "react-native-vector-icons/Entypo"
import CheckBox from "@react-native-community/checkbox"
import cc from "currency-codes"

import themeStyles from "./styles"
import {RowItem, RowSeparator} from "../../components/RowItem"
import connect from "./connect"
import {ThemeContext} from "../../ContextUtils/ThemeContext"
import {HeaderBar} from "../../components/HeaderBar"
import {ConversionContext} from "../../ContextUtils/ConversionContext"

// CurrencyList Component content
export const CurrencyList = () => {
  const {push} = useNavigation()
  const {name, params} = useRoute()

  const isBaseCurrency = get(params, "isBaseCurrency", false)
  const title = get(params, "title", "")

  const {styleableTheme} = useContext(ThemeContext)
  const styles = themeStyles(styleableTheme)

  const {
    baseCurrency,
    quoteCurrency,
    changeBaseCurrency,
    changeQuoteCurrency,
    rates,
    toggleFavorite,
  } = useContext(ConversionContext)

  const exclude = isBaseCurrency ? quoteCurrency : baseCurrency

  return (
    <View style={styles.root}>
      <HeaderBar title={title || name} />
      <FlatList
        data={filter(rates, (rate) => exclude !== rate.name)}
        renderItem={({item}) => {
          let selected = false

          if (isBaseCurrency && item.name === baseCurrency) {
            selected = true
          } else if (!isBaseCurrency && item.name === quoteCurrency) {
            selected = true
          }
          const {currency} = cc.code(item.name)
          return (
            <RowItem
              title={`${item.name} (${currency})`}
              isButton={false}
              rightIcon={
                <View
                  style={{
                    flexDirection: "row",
                  }}>
                  <TouchableOpacity
                    style={{
                      paddingRight: 12,
                    }}
                    onPress={() => toggleFavorite(item.name)}>
                    <Entypo
                      name={item.isFavorite ? "star" : "star-outlined"}
                      size={30}
                      color={item.isFavorite ? "#FFDE00" : styleableTheme[500]}
                    />
                  </TouchableOpacity>
                  <CheckBox
                    value={selected}
                    tintColor={styleableTheme[900]}
                    onCheckColor={styleableTheme[50]}
                    onFillColor={styleableTheme[700]}
                    onTintColor={styleableTheme[900]}
                    onValueChange={() => {
                      if (isBaseCurrency) {
                        changeBaseCurrency(item.name)
                      } else {
                        changeQuoteCurrency(item.name)
                      }
                      push("Home")
                    }}
                  />
                </View>
              }
            />
          )
        }}
        keyExtractor={(item) => item.name}
        ItemSeparatorComponent={() => <RowSeparator />}
        ListFooterComponent={() => <View />}
      />
    </View>
  )
}

// CurrencyList Proptypes
CurrencyList.propTypes = {}

// CurrencyList Default props
CurrencyList.defaultProps = {}

export default flow([connect])(CurrencyList)
