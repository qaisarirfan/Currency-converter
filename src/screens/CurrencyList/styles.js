import {StyleSheet} from "react-native"

// CurrencyList: Styling here
const styles = (styleableTheme) => {
  return StyleSheet.create({
    root: {
      flex: 1,
      backgroundColor: styleableTheme[500],
    },
  })
}

export default styles
