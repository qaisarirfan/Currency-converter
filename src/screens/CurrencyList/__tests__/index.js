import React from "react"
import {render, cleanup, store} from '../../../../jest/testUtils';
import CurrencyList from ".."

afterEach(() => {
  cleanup();
  store.clearActions();
});

describe("<CurrencyList />", () => {
  const createProps = () => ({})
  it("Does not explode", () => {
    const props = createProps()
    const component = render(<CurrencyList {...props} />).toJSON()
    expect(component).toMatchSnapshot()
  })
})
