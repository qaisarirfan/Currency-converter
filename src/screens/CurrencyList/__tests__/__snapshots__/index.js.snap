// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CurrencyList /> Does not explode 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#4f6d7a",
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {},
        Object {
          "backgroundColor": "#36505d",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "height": 20,
          },
          Object {
            "backgroundColor": "#36505d",
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "flexDirection": "row",
            "flexGrow": 1,
            "justifyContent": "space-between",
            "paddingBottom": 12,
            "paddingTop": 12,
          },
          Object {
            "backgroundColor": "#36505d",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "justifyContent": "center",
            "paddingLeft": 6,
            "width": 250,
          }
        }
      >
        <RNGestureHandlerButton
          collapsable={false}
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          rippleColor={0}
          testID="back"
        >
          <View
            accessible={true}
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "opacity": 1,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#eaedef",
                    "fontSize": 16,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Entypo",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              ï…‚
            </Text>
            <Text
              style={
                Object {
                  "color": "#eaedef",
                  "fontSize": 18,
                  "lineHeight": 18,
                }
              }
            >
              common.back
            </Text>
          </View>
        </RNGestureHandlerButton>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#36505d",
            "justifyContent": "center",
            "width": 250,
          }
        }
      >
        <Text
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "#eaedef",
              "fontSize": 18,
              "lineHeight": 18,
            }
          }
        >
          mockConstructor
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "justifyContent": "center",
            "paddingRight": 6,
            "width": 250,
          }
        }
      />
    </View>
  </View>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListFooterComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
